In [1]:
!pip install nltk

In [36]:
#
# Lastar inn teksten vi jobbar med
#

with open('./Input text - Chapter 3.txt', 'r') as f:
    file_contents = f.read()
    
print(file_contents)

3. Fire, machinery and electric installations
Section 7. Scope of application of this chapter
(1) This chapter applies to vessels constructed after 2 January 1988.
(2) Vessels constructed before 2 January 1988 shall be constructed and equipped in accordance with the rules
applicable at the time of construction of the vessel, unless otherwise provided by the individual provisions of
this chapter.
(3) When a vessel is substantially altered or modified, the vessel shall comply with the requirements for a new
vessel.
0 Amended by Regulation of 4 May 2015 No.. 445 (in force on 6 May 2015).
Section 8. Fire safety
(1) Vessels constructed after 1 January 1992 shall have structural fire protection in accordance with chapter
C14 of Nordic Boat Standard (1990), unless otherwise provided by the individual provisions of this chapter.
(2) Vessels constructed before 1 January 1992 shall have structural fire protection in accordance with the rules
applicable at the time of construction of the vessel.


In [38]:
#
# Splitte opp linjehopp frå tekst fil.
# Eksempel: Fire, machinery and electric installations\nSection 7.
#

forward_file_content_split = file_contents.split('\n')
forward_file_content_split

['3. Fire, machinery and electric installations',
 'Section 7. Scope of application of this chapter',
 '(1) This chapter applies to vessels constructed after 2 January 1988.',
 '(2) Vessels constructed before 2 January 1988 shall be constructed and equipped in accordance with the rules',
 'applicable at the time of construction of the vessel, unless otherwise provided by the individual provisions of',
 'this chapter.',
 '(3) When a vessel is substantially altered or modified, the vessel shall comply with the requirements for a new',
 'vessel.',
 '0 Amended by Regulation of 4 May 2015 No.. 445 (in force on 6 May 2015).',
 'Section 8. Fire safety',
 '(1) Vessels constructed after 1 January 1992 shall have structural fire protection in accordance with chapter',
 'C14 of Nordic Boat Standard (1990), unless otherwise provided by the individual provisions of this chapter.',
 '(2) Vessels constructed before 1 January 1992 shall have structural fire protection in accordance with the rules',
 '

In [39]:
#
# Transform list to one text
#

forward_one_text = " ".join(forward_file_content_split)
forward_one_text

"3. Fire, machinery and electric installations Section 7. Scope of application of this chapter (1) This chapter applies to vessels constructed after 2 January 1988. (2) Vessels constructed before 2 January 1988 shall be constructed and equipped in accordance with the rules applicable at the time of construction of the vessel, unless otherwise provided by the individual provisions of this chapter. (3) When a vessel is substantially altered or modified, the vessel shall comply with the requirements for a new vessel. 0 Amended by Regulation of 4 May 2015 No.. 445 (in force on 6 May 2015). Section 8. Fire safety (1) Vessels constructed after 1 January 1992 shall have structural fire protection in accordance with chapter C14 of Nordic Boat Standard (1990), unless otherwise provided by the individual provisions of this chapter. (2) Vessels constructed before 1 January 1992 shall have structural fire protection in accordance with the rules applicable at the time of construction of the vessel.

In [40]:
from nltk.tokenize import word_tokenize

forward_word_tokens = word_tokenize(forward_one_text, language="english")
forward_word_tokens

['3',
 '.',
 'Fire',
 ',',
 'machinery',
 'and',
 'electric',
 'installations',
 'Section',
 '7',
 '.',
 'Scope',
 'of',
 'application',
 'of',
 'this',
 'chapter',
 '(',
 '1',
 ')',
 'This',
 'chapter',
 'applies',
 'to',
 'vessels',
 'constructed',
 'after',
 '2',
 'January',
 '1988',
 '.',
 '(',
 '2',
 ')',
 'Vessels',
 'constructed',
 'before',
 '2',
 'January',
 '1988',
 'shall',
 'be',
 'constructed',
 'and',
 'equipped',
 'in',
 'accordance',
 'with',
 'the',
 'rules',
 'applicable',
 'at',
 'the',
 'time',
 'of',
 'construction',
 'of',
 'the',
 'vessel',
 ',',
 'unless',
 'otherwise',
 'provided',
 'by',
 'the',
 'individual',
 'provisions',
 'of',
 'this',
 'chapter',
 '.',
 '(',
 '3',
 ')',
 'When',
 'a',
 'vessel',
 'is',
 'substantially',
 'altered',
 'or',
 'modified',
 ',',
 'the',
 'vessel',
 'shall',
 'comply',
 'with',
 'the',
 'requirements',
 'for',
 'a',
 'new',
 'vessel',
 '.',
 '0',
 'Amended',
 'by',
 'Regulation',
 'of',
 '4',
 'May',
 '2015',
 'No',
 '..',
 '4

In [41]:
#
# Create sentences
#

from nltk.tokenize.punkt import PunktSentenceTokenizer

pst = PunktSentenceTokenizer()

forward_sentences = pst.sentences_from_tokens(forward_word_tokens)
list(forward_sentences)

[['3', '.'],
 ['Fire',
  ',',
  'machinery',
  'and',
  'electric',
  'installations',
  'Section',
  '7',
  '.'],
 ['Scope',
  'of',
  'application',
  'of',
  'this',
  'chapter',
  '(',
  '1',
  ')',
  'This',
  'chapter',
  'applies',
  'to',
  'vessels',
  'constructed',
  'after',
  '2',
  'January',
  '1988',
  '.'],
 ['(',
  '2',
  ')',
  'Vessels',
  'constructed',
  'before',
  '2',
  'January',
  '1988',
  'shall',
  'be',
  'constructed',
  'and',
  'equipped',
  'in',
  'accordance',
  'with',
  'the',
  'rules',
  'applicable',
  'at',
  'the',
  'time',
  'of',
  'construction',
  'of',
  'the',
  'vessel',
  ',',
  'unless',
  'otherwise',
  'provided',
  'by',
  'the',
  'individual',
  'provisions',
  'of',
  'this',
  'chapter',
  '.'],
 ['(',
  '3',
  ')',
  'When',
  'a',
  'vessel',
  'is',
  'substantially',
  'altered',
  'or',
  'modified',
  ',',
  'the',
  'vessel',
  'shall',
  'comply',
  'with',
  'the',
  'requirements',
  'for',
  'a',
  'new',
  'vessel

In [42]:
# Splitte opp tekst etter setningar.
# Vi leitar derfor etter ord som høyrer saman innanfor ein setning.

forward_sent_tokens = pst.tokenize(forward_one_text)
forward_sent_tokens

['3.',
 'Fire, machinery and electric installations Section 7.',
 'Scope of application of this chapter (1) This chapter applies to vessels constructed after 2 January 1988.',
 '(2) Vessels constructed before 2 January 1988 shall be constructed and equipped in accordance with the rules applicable at the time of construction of the vessel, unless otherwise provided by the individual provisions of this chapter.',
 '(3) When a vessel is substantially altered or modified, the vessel shall comply with the requirements for a new vessel.',
 '0 Amended by Regulation of 4 May 2015 No.. 445 (in force on 6 May 2015).',
 'Section 8.',
 'Fire safety (1) Vessels constructed after 1 January 1992 shall have structural fire protection in accordance with chapter C14 of Nordic Boat Standard (1990), unless otherwise provided by the individual provisions of this chapter.',
 '(2) Vessels constructed before 1 January 1992 shall have structural fire protection in accordance with the rules applicable at the ti

In [43]:
#
# Each line has a sentence index number.
# It is important that we keep these sentence index number.
# It will be the key to get a result in the end.
#

for sentence_index_number,sentence_as_a_string in enumerate(forward_sent_tokens):
    print("--- sentence_index_number: " + str(sentence_index_number) + " ---")
    print(sentence_as_a_string)

--- sentence_index_number: 0 ---
3.
--- sentence_index_number: 1 ---
Fire, machinery and electric installations Section 7.
--- sentence_index_number: 2 ---
Scope of application of this chapter (1) This chapter applies to vessels constructed after 2 January 1988.
--- sentence_index_number: 3 ---
(2) Vessels constructed before 2 January 1988 shall be constructed and equipped in accordance with the rules applicable at the time of construction of the vessel, unless otherwise provided by the individual provisions of this chapter.
--- sentence_index_number: 4 ---
(3) When a vessel is substantially altered or modified, the vessel shall comply with the requirements for a new vessel.
--- sentence_index_number: 5 ---
0 Amended by Regulation of 4 May 2015 No.. 445 (in force on 6 May 2015).
--- sentence_index_number: 6 ---
Section 8.
--- sentence_index_number: 7 ---
Fire safety (1) Vessels constructed after 1 January 1992 shall have structural fire protection in accordance with chapter C14 of Nord

In [75]:
#
# Make line ready for POS creation
#

lines_ready_for_pos_creation = []

for sentence_index_number,sentence_as_a_string in enumerate(forward_sent_tokens):
    
    tokens = word_tokenize(sentence_as_a_string)
    
    lines_ready_for_pos_creation.append(tokens)
    
lines_ready_for_pos_creation

[['3', '.'],
 ['Fire',
  ',',
  'machinery',
  'and',
  'electric',
  'installations',
  'Section',
  '7',
  '.'],
 ['Scope',
  'of',
  'application',
  'of',
  'this',
  'chapter',
  '(',
  '1',
  ')',
  'This',
  'chapter',
  'applies',
  'to',
  'vessels',
  'constructed',
  'after',
  '2',
  'January',
  '1988',
  '.'],
 ['(',
  '2',
  ')',
  'Vessels',
  'constructed',
  'before',
  '2',
  'January',
  '1988',
  'shall',
  'be',
  'constructed',
  'and',
  'equipped',
  'in',
  'accordance',
  'with',
  'the',
  'rules',
  'applicable',
  'at',
  'the',
  'time',
  'of',
  'construction',
  'of',
  'the',
  'vessel',
  ',',
  'unless',
  'otherwise',
  'provided',
  'by',
  'the',
  'individual',
  'provisions',
  'of',
  'this',
  'chapter',
  '.'],
 ['(',
  '3',
  ')',
  'When',
  'a',
  'vessel',
  'is',
  'substantially',
  'altered',
  'or',
  'modified',
  ',',
  'the',
  'vessel',
  'shall',
  'comply',
  'with',
  'the',
  'requirements',
  'for',
  'a',
  'new',
  'vessel

In [76]:
# Get the type of the words

forward_sentences_in_word_list_with_pos = []

for token_line in lines_ready_for_pos_creation:
    tagged_words = nltk.pos_tag(token_line)
    forward_sentences_in_word_list_with_pos.append(tagged_words)
    
forward_sentences_in_word_list_with_pos

[[('3', 'CD'), ('.', '.')],
 [('Fire', 'NNP'),
  (',', ','),
  ('machinery', 'NN'),
  ('and', 'CC'),
  ('electric', 'JJ'),
  ('installations', 'NNS'),
  ('Section', 'NN'),
  ('7', 'CD'),
  ('.', '.')],
 [('Scope', 'NN'),
  ('of', 'IN'),
  ('application', 'NN'),
  ('of', 'IN'),
  ('this', 'DT'),
  ('chapter', 'NN'),
  ('(', '('),
  ('1', 'CD'),
  (')', ')'),
  ('This', 'DT'),
  ('chapter', 'NN'),
  ('applies', 'VBZ'),
  ('to', 'TO'),
  ('vessels', 'NNS'),
  ('constructed', 'VBN'),
  ('after', 'IN'),
  ('2', 'CD'),
  ('January', 'NNP'),
  ('1988', 'CD'),
  ('.', '.')],
 [('(', '('),
  ('2', 'CD'),
  (')', ')'),
  ('Vessels', 'NNS'),
  ('constructed', 'VBN'),
  ('before', 'IN'),
  ('2', 'CD'),
  ('January', 'NNP'),
  ('1988', 'CD'),
  ('shall', 'MD'),
  ('be', 'VB'),
  ('constructed', 'VBN'),
  ('and', 'CC'),
  ('equipped', 'VBN'),
  ('in', 'IN'),
  ('accordance', 'NN'),
  ('with', 'IN'),
  ('the', 'DT'),
  ('rules', 'NNS'),
  ('applicable', 'VBP'),
  ('at', 'IN'),
  ('the', 'DT'),
  ('ti

In [79]:
#
# Print bigram on all text
#

from nltk import ngrams

for sentence_index_number,sentence_as_a_string in enumerate(forward_sent_tokens):
    
    token = nltk.word_tokenize(sentence_as_a_string)
    tagged_words = nltk.pos_tag(token)
    
    print(token)
    print(tagged_words)
    #bigram = list(ngrams(token, 2))
    #print(bigram)

['3', '.']
[('3', 'CD'), ('.', '.')]
['Fire', ',', 'machinery', 'and', 'electric', 'installations', 'Section', '7', '.']
[('Fire', 'NNP'), (',', ','), ('machinery', 'NN'), ('and', 'CC'), ('electric', 'JJ'), ('installations', 'NNS'), ('Section', 'NN'), ('7', 'CD'), ('.', '.')]
['Scope', 'of', 'application', 'of', 'this', 'chapter', '(', '1', ')', 'This', 'chapter', 'applies', 'to', 'vessels', 'constructed', 'after', '2', 'January', '1988', '.']
[('Scope', 'NN'), ('of', 'IN'), ('application', 'NN'), ('of', 'IN'), ('this', 'DT'), ('chapter', 'NN'), ('(', '('), ('1', 'CD'), (')', ')'), ('This', 'DT'), ('chapter', 'NN'), ('applies', 'VBZ'), ('to', 'TO'), ('vessels', 'NNS'), ('constructed', 'VBN'), ('after', 'IN'), ('2', 'CD'), ('January', 'NNP'), ('1988', 'CD'), ('.', '.')]
['(', '2', ')', 'Vessels', 'constructed', 'before', '2', 'January', '1988', 'shall', 'be', 'constructed', 'and', 'equipped', 'in', 'accordance', 'with', 'the', 'rules', 'applicable', 'at', 'the', 'time', 'of', 'construct

['(', '7', ')', 'Stoves', 'shall', 'have', 'smoke', 'outlet', 'to', 'uptake', '.']
[('(', '('), ('7', 'CD'), (')', ')'), ('Stoves', 'NNPS'), ('shall', 'MD'), ('have', 'VB'), ('smoke', 'VBN'), ('outlet', 'NN'), ('to', 'TO'), ('uptake', 'VB'), ('.', '.')]
['Section', '10', '.']
[('Section', 'NN'), ('10', 'CD'), ('.', '.')]
['Installation', ',', 'etc', '.']
[('Installation', 'NN'), (',', ','), ('etc', 'FW'), ('.', '.')]
['of', 'oilfired', 'stove', '(', '1', ')', 'Stoves', 'shall', 'be', 'securely', 'fixed', 'to', 'the', 'floor', 'or', 'to', 'a', 'fixed', 'bench', ',', 'if', 'possible', 'near', 'the', 'centre', 'line', 'of', 'the', 'vessel', '.']
[('of', 'IN'), ('oilfired', 'VBN'), ('stove', 'NN'), ('(', '('), ('1', 'CD'), (')', ')'), ('Stoves', 'VBZ'), ('shall', 'MD'), ('be', 'VB'), ('securely', 'RB'), ('fixed', 'VBN'), ('to', 'TO'), ('the', 'DT'), ('floor', 'NN'), ('or', 'CC'), ('to', 'TO'), ('a', 'DT'), ('fixed', 'VBN'), ('bench', 'NN'), (',', ','), ('if', 'IN'), ('possible', 'JJ'), ('n

['Smoke', 'uptake', 'and', 'exhaust', 'pipes', '(', '1', ')', 'Smoke', 'uptakes', 'and', 'exhaust', 'pipes', 'shall', 'consist', 'of', 'at', 'least', '3', 'mm', 'thick', 'steel', 'plates', 'and', 'shall', 'be', 'securely', 'fixed', '.']
[('Smoke', 'NNP'), ('uptake', 'NN'), ('and', 'CC'), ('exhaust', 'JJ'), ('pipes', 'NNS'), ('(', '('), ('1', 'CD'), (')', ')'), ('Smoke', 'NNP'), ('uptakes', 'NNS'), ('and', 'CC'), ('exhaust', 'JJ'), ('pipes', 'NNS'), ('shall', 'MD'), ('consist', 'VB'), ('of', 'IN'), ('at', 'IN'), ('least', 'JJS'), ('3', 'CD'), ('mm', 'JJ'), ('thick', 'NN'), ('steel', 'NN'), ('plates', 'NNS'), ('and', 'CC'), ('shall', 'MD'), ('be', 'VB'), ('securely', 'RB'), ('fixed', 'VBN'), ('.', '.')]
['Smoke', 'uptakes', 'shall', 'be', 'carried', 'at', 'least', '1.3', 'm', 'above', 'deck', '.']
[('Smoke', 'NNP'), ('uptakes', 'NNS'), ('shall', 'MD'), ('be', 'VB'), ('carried', 'VBN'), ('at', 'IN'), ('least', 'JJS'), ('1.3', 'CD'), ('m', 'NNS'), ('above', 'IN'), ('deck', 'NN'), ('.', '.'

['shall', 'be', 'insulated', 'with', 'non-combustible', 'material', '.']
[('shall', 'MD'), ('be', 'VB'), ('insulated', 'VBN'), ('with', 'IN'), ('non-combustible', 'JJ'), ('material', 'NN'), ('.', '.')]
['When', 'an', 'engine', 'cover', 'is', 'used', ',', 'this', 'shall', 'be', 'insulated', 'on', 'the', 'inside', 'and', 'be', 'mounted', 'with', 'a', 'galvanized', 'steel', 'or', 'aluminium', 'sheet', '.']
[('When', 'WRB'), ('an', 'DT'), ('engine', 'NN'), ('cover', 'NN'), ('is', 'VBZ'), ('used', 'VBN'), (',', ','), ('this', 'DT'), ('shall', 'MD'), ('be', 'VB'), ('insulated', 'VBN'), ('on', 'IN'), ('the', 'DT'), ('inside', 'NN'), ('and', 'CC'), ('be', 'VB'), ('mounted', 'VBN'), ('with', 'IN'), ('a', 'DT'), ('galvanized', 'JJ'), ('steel', 'NN'), ('or', 'CC'), ('aluminium', 'JJ'), ('sheet', 'NN'), ('.', '.')]
['Section', '15', '.']
[('Section', 'NN'), ('15', 'CD'), ('.', '.')]
['Fire-extinguishing', 'systems', 'in', 'machinery', 'spaces', '(', '1', ')', 'Vessels', 'of', '10.67', 'to', '15', 

['For', 'vessels', 'without', 'wheelhouse', 'or', 'accommodation', 'spaces', 'one', 'fire', 'extinguisher', 'is', 'sufficient', '.']
[('For', 'IN'), ('vessels', 'NNS'), ('without', 'IN'), ('wheelhouse', 'NN'), ('or', 'CC'), ('accommodation', 'NN'), ('spaces', 'NNS'), ('one', 'CD'), ('fire', 'NN'), ('extinguisher', 'NN'), ('is', 'VBZ'), ('sufficient', 'JJ'), ('.', '.')]
['(', '1', ')', 'Fire', 'extinguishers', 'with', 'pressurised', 'extinguishing', 'medium', 'shall', 'not', 'be', 'placed', 'in', 'accommodation', 'spaces', '.']
[('(', '('), ('1', 'CD'), (')', ')'), ('Fire', 'NN'), ('extinguishers', 'NNS'), ('with', 'IN'), ('pressurised', 'JJ'), ('extinguishing', 'VBG'), ('medium', 'NN'), ('shall', 'MD'), ('not', 'RB'), ('be', 'VB'), ('placed', 'VBN'), ('in', 'IN'), ('accommodation', 'NN'), ('spaces', 'NNS'), ('.', '.')]
['Fire', 'extinguishers', 'which', 'may', 'be', 'exposed', 'to', 'frost', 'shall', 'be', 'of', 'a', 'frost-proof', 'type', '.']
[('Fire', 'NN'), ('extinguishers', 'NNS')

['Machinery', 'and', 'piping', 'systems', '(', '1', ')', 'Vessels', 'constructed', 'after', '1', 'January', '1992', 'shall', 'satisfy', 'the', 'requirements', 'of', 'Nordic', 'Boat', 'Standard', '(', '1990', ')', 'a', ')', 'C6', 'on', 'bilge', 'systems', ';', 'b', ')', 'C9', 'on', 'oil', 'fuel', 'systems', ';', 'c', ')', 'C8', 'on', 'engine', 'installations', ';', 'and', 'd', ')', 'C7', 'and', 'C10', 'on', 'rudder', 'and', 'propeller', 'arrangements', '.']
[('Machinery', 'NN'), ('and', 'CC'), ('piping', 'VBG'), ('systems', 'NNS'), ('(', '('), ('1', 'CD'), (')', ')'), ('Vessels', 'NNS'), ('constructed', 'VBN'), ('after', 'IN'), ('1', 'CD'), ('January', 'NNP'), ('1992', 'CD'), ('shall', 'MD'), ('satisfy', 'VB'), ('the', 'DT'), ('requirements', 'NNS'), ('of', 'IN'), ('Nordic', 'NNP'), ('Boat', 'NNP'), ('Standard', 'NNP'), ('(', '('), ('1990', 'CD'), (')', ')'), ('a', 'DT'), (')', ')'), ('C6', 'NN'), ('on', 'IN'), ('bilge', 'JJ'), ('systems', 'NNS'), (';', ':'), ('b', 'NN'), (')', ')'), ('

['If', 'the', 'emergency', 'source', 'of', 'power', 'is', 'placed', 'in', 'the', 'machinery', 'space', ',', 'it', 'shall', 'be', 'placed', 'above', 'the', 'floor', 'and', 'in', 'a', 'well-ventilated', 'place', '.']
[('If', 'IN'), ('the', 'DT'), ('emergency', 'NN'), ('source', 'NN'), ('of', 'IN'), ('power', 'NN'), ('is', 'VBZ'), ('placed', 'VBN'), ('in', 'IN'), ('the', 'DT'), ('machinery', 'NN'), ('space', 'NN'), (',', ','), ('it', 'PRP'), ('shall', 'MD'), ('be', 'VB'), ('placed', 'VBN'), ('above', 'IN'), ('the', 'DT'), ('floor', 'NN'), ('and', 'CC'), ('in', 'IN'), ('a', 'DT'), ('well-ventilated', 'JJ'), ('place', 'NN'), ('.', '.')]
['(', '5', ')', 'The', 'emergency', 'source', 'of', 'power', 'and', 'the', 'emergency', 'lighting', 'on', 'vessels', 'constructed', 'before', '1', 'January', '1992', 'shall', 'be', 'in', 'accordance', 'with', 'the', 'rules', 'in', 'force', 'at', 'the', 'time', 'of', 'construction', '.']
[('(', '('), ('5', 'CD'), (')', ')'), ('The', 'DT'), ('emergency', 'NN')

In [94]:
#
# Print bigram on all text
# Chalange: Can we keep the POS through the bigram process? This can be very useful! When lemmatizing..
#

from nltk import ngrams

forward_bigrams_with_pos_preserved = []

for sentence_index_number,sentence_as_a_string in enumerate(forward_sent_tokens):
    
    tokenized_sentence = nltk.word_tokenize(sentence_as_a_string)
    sentence_pos_tagged_by_word_list = nltk.pos_tag(tokenized_sentence)
    
    sentence_as_pos = []
    
    for word_index_number,tokenized_word in enumerate(tokenized_sentence):
        #print(tokenized_sentence[word_index_number]) # the word
        #print(sentence_pos_tagged_by_word_list[word_index_number][1]) # POS of the word
        sentence_as_pos.append(sentence_pos_tagged_by_word_list[word_index_number][1])
    
    # Now these two are the same, but one is the POS
    # Idea: We run bigram on both to keep the POS
    #print(tokenized_sentence)
    #print(sentence_as_pos)
    
    bigram_on_tokenized_sentence = list(ngrams(tokenized_sentence, 2))
    bigram_on_sentence_as_pos = list(ngrams(sentence_as_pos, 2))
    #print(bigram_on_tokenized_sentence)
    #print(bigram_on_sentence_as_pos)
    
    # We did it!!
    
    # Now we merge the results to one list
    bigram_on_tokenized_sentence_with_pos_preserved = (bigram_on_tokenized_sentence,bigram_on_sentence_as_pos)
    forward_bigrams_with_pos_preserved.append(bigram_on_tokenized_sentence_with_pos_preserved)
    
forward_bigrams_with_pos_preserved

[([('3', '.')], [('CD', '.')]),
 ([('Fire', ','),
   (',', 'machinery'),
   ('machinery', 'and'),
   ('and', 'electric'),
   ('electric', 'installations'),
   ('installations', 'Section'),
   ('Section', '7'),
   ('7', '.')],
  [('NNP', ','),
   (',', 'NN'),
   ('NN', 'CC'),
   ('CC', 'JJ'),
   ('JJ', 'NNS'),
   ('NNS', 'NN'),
   ('NN', 'CD'),
   ('CD', '.')]),
 ([('Scope', 'of'),
   ('of', 'application'),
   ('application', 'of'),
   ('of', 'this'),
   ('this', 'chapter'),
   ('chapter', '('),
   ('(', '1'),
   ('1', ')'),
   (')', 'This'),
   ('This', 'chapter'),
   ('chapter', 'applies'),
   ('applies', 'to'),
   ('to', 'vessels'),
   ('vessels', 'constructed'),
   ('constructed', 'after'),
   ('after', '2'),
   ('2', 'January'),
   ('January', '1988'),
   ('1988', '.')],
  [('NN', 'IN'),
   ('IN', 'NN'),
   ('NN', 'IN'),
   ('IN', 'DT'),
   ('DT', 'NN'),
   ('NN', '('),
   ('(', 'CD'),
   ('CD', ')'),
   (')', 'DT'),
   ('DT', 'NN'),
   ('NN', 'VBZ'),
   ('VBZ', 'TO'),
   ('TO', 'N

In [95]:
#
# Get the stop words
#

import nltk
from nltk.corpus import stopwords
nltk.download('stopwords')

forward_english_stopwords = set(stopwords.words('english'))

# Legger til alt som ikkje er til hjelp for å finne dobbeltord
forward_english_stopwords.update(['.',',','(',')','c',';','..',':','b','c','the','in',"'s",'no','an','etc','where','-'])

# Removing the months
forward_english_stopwords.update(['january','february','march','april','may','june','july','august','september','october','november','december'])

forward_english_stopwords

[nltk_data] Downloading package stopwords to /home/jovyan/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


{"'s",
 '(',
 ')',
 ',',
 '-',
 '.',
 '..',
 ':',
 ';',
 'a',
 'about',
 'above',
 'after',
 'again',
 'against',
 'ain',
 'all',
 'am',
 'an',
 'and',
 'any',
 'april',
 'are',
 'aren',
 "aren't",
 'as',
 'at',
 'august',
 'b',
 'be',
 'because',
 'been',
 'before',
 'being',
 'below',
 'between',
 'both',
 'but',
 'by',
 'c',
 'can',
 'couldn',
 "couldn't",
 'd',
 'december',
 'did',
 'didn',
 "didn't",
 'do',
 'does',
 'doesn',
 "doesn't",
 'doing',
 'don',
 "don't",
 'down',
 'during',
 'each',
 'etc',
 'february',
 'few',
 'for',
 'from',
 'further',
 'had',
 'hadn',
 "hadn't",
 'has',
 'hasn',
 "hasn't",
 'have',
 'haven',
 "haven't",
 'having',
 'he',
 'her',
 'here',
 'hers',
 'herself',
 'him',
 'himself',
 'his',
 'how',
 'i',
 'if',
 'in',
 'into',
 'is',
 'isn',
 "isn't",
 'it',
 "it's",
 'its',
 'itself',
 'january',
 'july',
 'june',
 'just',
 'll',
 'm',
 'ma',
 'march',
 'may',
 'me',
 'mightn',
 "mightn't",
 'more',
 'most',
 'mustn',
 "mustn't",
 'my',
 'myself',
 'ne

In [64]:
# Does the input string contain a digit?
def hasNumbers(inputString):
    return any(char.isdigit() for char in inputString)

def hasSlash(inputString):
    return any(char == '/' for char in inputString)

In [65]:
#
# Print bigram with no stopwords
#

from nltk import ngrams

for sentence_index_number,sentence_as_a_string in enumerate(forward_sent_tokens):
    token = nltk.word_tokenize(sentence_as_a_string)
    bigram_list = list(ngrams(token, 2))
    
    for bigram in bigram_list:
        
        if hasNumbers(bigram[0]) or hasNumbers(bigram[1]):
            continue # skip if containing numbers
        elif hasSlash(bigram[0]) or hasSlash(bigram[1]):
            continue # skip if containing /
        
        bigram_in_a_list = [bigram[0],bigram[1]]
        
        bigram_without_stopword = [bigram_split for bigram_split in bigram_in_a_list if bigram_split.lower() not in forward_english_stopwords]
        
        if len(bigram_without_stopword) == 2:
            print(bigram_without_stopword)
        

['electric', 'installations']
['installations', 'Section']
['chapter', 'applies']
['vessels', 'constructed']
['Vessels', 'constructed']
['rules', 'applicable']
['unless', 'otherwise']
['otherwise', 'provided']
['individual', 'provisions']
['substantially', 'altered']
['vessel', 'shall']
['shall', 'comply']
['new', 'vessel']
['Fire', 'safety']
['Vessels', 'constructed']
['structural', 'fire']
['fire', 'protection']
['Nordic', 'Boat']
['Boat', 'Standard']
['unless', 'otherwise']
['otherwise', 'provided']
['individual', 'provisions']
['Vessels', 'constructed']
['structural', 'fire']
['fire', 'protection']
['rules', 'applicable']
['gasfired', 'installations']
['installations', 'shall']
['shall', 'satisfy']
['safety', 'precautions']
['gasfired', 'installations']
['liquefied', 'hydrocarbon']
['hydrocarbon', 'gases']
['gases', 'used']
['board', 'vessels']
['Oil-fired', 'stoves']
['electric', 'power']
['power', 'supply']
['supply', 'shall']
['oil', 'supply']
['smoke', 'gas']
['flame', 'goes']


# Remove all the words we don't need

In [44]:
#
# Clean the lines up
#

import re

lines = []

for line in forward_sent_tokens:
    
    # remove all numbers
    edit_1 = re.sub(r"\d","", line)
    
    edit_2 = edit_1.replace('()','')
    
    edit_3 = edit_2.replace('/',' ') # eksempel: splitte sånt som dette i to -> 'space/locker'
    
    edit_4 = edit_3.replace('Section .','')
    
    edit_5 = edit_4.replace(' m ',' metre ')
    
    edit_6 = edit_5.replace(' V ',' voltage ')
    
    # remove spaces at the beginning and at the end of the string
    edit_7 = edit_6.strip()
    
    if edit_7 != '':
        lines.append(edit_7)
    
lines

['.',
 'Fire, machinery and electric installations',
 'Scope of application of this chapter  This chapter applies to vessels constructed after  January .',
 'Vessels constructed before  January  shall be constructed and equipped in accordance with the rules applicable at the time of construction of the vessel, unless otherwise provided by the individual provisions of this chapter.',
 'When a vessel is substantially altered or modified, the vessel shall comply with the requirements for a new vessel.',
 'Amended by Regulation of  May  No..  (in force on  May ).',
 'Fire safety  Vessels constructed after  January  shall have structural fire protection in accordance with chapter C of Nordic Boat Standard , unless otherwise provided by the individual provisions of this chapter.',
 'Vessels constructed before  January  shall have structural fire protection in accordance with the rules applicable at the time of construction of the vessel.',
 'Irrespective of the year of construction of the ve

In [10]:
#
# Removing stop words
#

stopped_tokens = []

new_lines = []
new_lines_as_text = []

for line in lines:
    
    tokens = word_tokenize(line)
    
    stopped_tokens = stopped_tokens + [token for token in tokens if token.lower() in stop_words]
    
    new_line_in_the_making = [token for token in tokens if token.lower() not in stop_words]
    new_line = " ".join(new_line_in_the_making)
    
    # only add lines that are not blank
    if new_line != '':
        new_lines.append(new_line_in_the_making)
        new_lines_as_text.append(new_line)
    
#print(stopped_tokens)
#print(new_lines)
print(new_lines_as_text)

['Fire machinery electric installations', 'Scope application chapter chapter applies vessels constructed', 'Vessels constructed shall constructed equipped accordance rules applicable time construction vessel unless otherwise provided individual provisions chapter', 'vessel substantially altered modified vessel shall comply requirements new vessel', 'Amended Regulation force', 'Fire safety Vessels constructed shall structural fire protection accordance chapter Nordic Boat Standard unless otherwise provided individual provisions chapter', 'Vessels constructed shall structural fire protection accordance rules applicable time construction vessel', 'Irrespective year construction vessel gasfired installations shall satisfy requirements Regulations', 'safety precautions gasfired installations', 'operating propane liquefied hydrocarbon gases used board vessels', 'Oil-fired stoves Stoves electric power supply shall oil supply shuts automatically stove over-heated maximum °C smoke gas flame goe

In [11]:
#
# From here we are done changing the text.
# The words will have the same index, even though we have a text with lower case and one normal case text.
#

In [12]:
# Get the type of the words

lines_with_word_type = []

for token_line in new_lines:
    tagged_words = nltk.pos_tag(token_line)
    lines_with_word_type.append(tagged_words)
    
lines_with_word_type

[[('Fire', 'NNP'),
  ('machinery', 'NN'),
  ('electric', 'JJ'),
  ('installations', 'NNS')],
 [('Scope', 'NNP'),
  ('application', 'NN'),
  ('chapter', 'NN'),
  ('chapter', 'NN'),
  ('applies', 'NNS'),
  ('vessels', 'NNS'),
  ('constructed', 'VBD')],
 [('Vessels', 'NNS'),
  ('constructed', 'VBN'),
  ('shall', 'MD'),
  ('constructed', 'VBN'),
  ('equipped', 'JJ'),
  ('accordance', 'NN'),
  ('rules', 'NNS'),
  ('applicable', 'JJ'),
  ('time', 'NN'),
  ('construction', 'NN'),
  ('vessel', 'NN'),
  ('unless', 'IN'),
  ('otherwise', 'RB'),
  ('provided', 'VBN'),
  ('individual', 'JJ'),
  ('provisions', 'NNS'),
  ('chapter', 'NN')],
 [('vessel', 'NN'),
  ('substantially', 'RB'),
  ('altered', 'VBD'),
  ('modified', 'JJ'),
  ('vessel', 'NN'),
  ('shall', 'MD'),
  ('comply', 'VB'),
  ('requirements', 'NNS'),
  ('new', 'JJ'),
  ('vessel', 'NN')],
 [('Amended', 'VBN'), ('Regulation', 'NNP'), ('force', 'NN')],
 [('Fire', 'NNP'),
  ('safety', 'NN'),
  ('Vessels', 'NNP'),
  ('constructed', 'VBD'),


In [13]:
#
# fjern bøying
#

from nltk.corpus import wordnet
from nltk.stem.wordnet import WordNetLemmatizer
nltk.download('wordnet')

def get_wordnet_pos(treebank_tag):

    if treebank_tag.startswith('J'):
        return wordnet.ADJ
    elif treebank_tag.startswith('V'):
        return wordnet.VERB
    elif treebank_tag.startswith('N'):
        return wordnet.NOUN
    elif treebank_tag.startswith('R'):
        return wordnet.ADV
    else:
        return ''

lines_with_lemmatized_words_and_tag = []
    
for token_line in lines_with_word_type:
    
    lemmatizer = WordNetLemmatizer()
    lemmatized_single_words_with_tag = []

    for (word, pos) in token_line:
        wordnet_pos = get_wordnet_pos(pos)
        if wordnet_pos != '':
            
            # Auto lemmatizer
            lemmatized_word = lemmatizer.lemmatize(word.lower(), wordnet_pos)
            
            # Custom lemmatizer
            if lemmatized_word == 'metres':
                lemmatized_word = 'metre' # (Commonwealth spelling)
            elif lemmatized_word == 'meter': # (American spelling)
                lemmatized_word = 'metre'
            
            lemmatized_single_words_with_tag.append((lemmatized_word, pos))
    
    lines_with_lemmatized_words_and_tag.append(lemmatized_single_words_with_tag)
    
lines_with_lemmatized_words_and_tag

[nltk_data] Downloading package wordnet to /home/jovyan/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


[[('fire', 'NNP'),
  ('machinery', 'NN'),
  ('electric', 'JJ'),
  ('installation', 'NNS')],
 [('scope', 'NNP'),
  ('application', 'NN'),
  ('chapter', 'NN'),
  ('chapter', 'NN'),
  ('applies', 'NNS'),
  ('vessel', 'NNS'),
  ('construct', 'VBD')],
 [('vessel', 'NNS'),
  ('construct', 'VBN'),
  ('construct', 'VBN'),
  ('equipped', 'JJ'),
  ('accordance', 'NN'),
  ('rule', 'NNS'),
  ('applicable', 'JJ'),
  ('time', 'NN'),
  ('construction', 'NN'),
  ('vessel', 'NN'),
  ('otherwise', 'RB'),
  ('provide', 'VBN'),
  ('individual', 'JJ'),
  ('provision', 'NNS'),
  ('chapter', 'NN')],
 [('vessel', 'NN'),
  ('substantially', 'RB'),
  ('alter', 'VBD'),
  ('modified', 'JJ'),
  ('vessel', 'NN'),
  ('comply', 'VB'),
  ('requirement', 'NNS'),
  ('new', 'JJ'),
  ('vessel', 'NN')],
 [('amend', 'VBN'), ('regulation', 'NNP'), ('force', 'NN')],
 [('fire', 'NNP'),
  ('safety', 'NN'),
  ('vessel', 'NNP'),
  ('construct', 'VBD'),
  ('structural', 'JJ'),
  ('fire', 'NN'),
  ('protection', 'NN'),
  ('accordan

# From this point the text will not be adjusted

In [14]:
#
# * Creating index on sentences in text
# * Creating index on words in text
# * Creating sentence span with the word index 
#

span_count = -1
line_index_with_span_for_lookup = []

for line_idx,line in enumerate(lines_with_lemmatized_words_and_tag):
    
    number_of_words_in_line = len(line)
    first_word_in_sentence_has_index = span_count + 1
    last_word_in_sentence_has_index = first_word_in_sentence_has_index + number_of_words_in_line - 1
    
    # update span counter
    span_count = last_word_in_sentence_has_index
    
    line_index_with_span_for_lookup.append((line_idx,first_word_in_sentence_has_index,last_word_in_sentence_has_index))
    
    print("Line index: " + str(line_idx) + " (" + str(first_word_in_sentence_has_index) + " - " + str(last_word_in_sentence_has_index) + ")")

line_index_with_span_for_lookup

Line index: 0 (0 - 3)
Line index: 1 (4 - 10)
Line index: 2 (11 - 25)
Line index: 3 (26 - 34)
Line index: 4 (35 - 37)
Line index: 5 (38 - 54)
Line index: 6 (55 - 65)
Line index: 7 (66 - 74)
Line index: 8 (75 - 78)
Line index: 9 (79 - 86)
Line index: 10 (87 - 109)
Line index: 11 (110 - 125)
Line index: 12 (126 - 145)
Line index: 13 (146 - 156)
Line index: 14 (157 - 167)
Line index: 15 (168 - 174)
Line index: 16 (175 - 178)
Line index: 17 (179 - 182)
Line index: 18 (183 - 186)
Line index: 19 (187 - 193)
Line index: 20 (194 - 197)
Line index: 21 (198 - 198)
Line index: 22 (199 - 210)
Line index: 23 (211 - 224)
Line index: 24 (225 - 233)
Line index: 25 (234 - 247)
Line index: 26 (248 - 254)
Line index: 27 (255 - 257)
Line index: 28 (258 - 265)
Line index: 29 (266 - 272)
Line index: 30 (273 - 279)
Line index: 31 (280 - 286)
Line index: 32 (287 - 298)
Line index: 33 (299 - 313)
Line index: 34 (314 - 317)
Line index: 35 (318 - 327)
Line index: 36 (328 - 329)
Line index: 37 (330 - 335)
Line ind

[(0, 0, 3),
 (1, 4, 10),
 (2, 11, 25),
 (3, 26, 34),
 (4, 35, 37),
 (5, 38, 54),
 (6, 55, 65),
 (7, 66, 74),
 (8, 75, 78),
 (9, 79, 86),
 (10, 87, 109),
 (11, 110, 125),
 (12, 126, 145),
 (13, 146, 156),
 (14, 157, 167),
 (15, 168, 174),
 (16, 175, 178),
 (17, 179, 182),
 (18, 183, 186),
 (19, 187, 193),
 (20, 194, 197),
 (21, 198, 198),
 (22, 199, 210),
 (23, 211, 224),
 (24, 225, 233),
 (25, 234, 247),
 (26, 248, 254),
 (27, 255, 257),
 (28, 258, 265),
 (29, 266, 272),
 (30, 273, 279),
 (31, 280, 286),
 (32, 287, 298),
 (33, 299, 313),
 (34, 314, 317),
 (35, 318, 327),
 (36, 328, 329),
 (37, 330, 335),
 (38, 336, 356),
 (39, 357, 364),
 (40, 365, 380),
 (41, 381, 386),
 (42, 387, 398),
 (43, 399, 408),
 (44, 409, 425),
 (45, 426, 440),
 (46, 441, 448),
 (47, 449, 456),
 (48, 457, 463),
 (49, 464, 481),
 (50, 482, 492),
 (51, 493, 496),
 (52, 497, 502),
 (53, 503, 513),
 (54, 514, 531),
 (55, 532, 542),
 (56, 543, 555),
 (57, 556, 563),
 (58, 564, 580),
 (59, 581, 590),
 (60, 591, 599

In [15]:
#
# word_index_for_lookup
# output, for each line:
# * index on word in text
# * index on sentence in text
# * word in text
# * POS tagging (part of speech) of the word
#

span_count = -1
line_index_with_span_for_lookup = []
word_index_for_lookup = []

for line_idx,line in enumerate(lines_with_lemmatized_words_and_tag):
    
    number_of_words_in_line = len(line)
    first_word_in_sentence_has_index = span_count + 1
    last_word_in_sentence_has_index = first_word_in_sentence_has_index + number_of_words_in_line - 1
    
    # update span counter
    span_count = last_word_in_sentence_has_index
    
    line_index_with_span_for_lookup.append((line_idx,first_word_in_sentence_has_index,last_word_in_sentence_has_index))
    
    word_index_count = first_word_in_sentence_has_index
    
    for (word_in_text,pos) in line:
        
        word_with_meta_info = (word_index_count,line_idx,word_in_text,pos)
        word_index_for_lookup.append(word_with_meta_info)
        
        # update counter
        word_index_count = word_index_count + 1
    
    #print("Line index: " + str(line_idx) + " (" + str(first_word_in_sentence_has_index) + " - " + str(last_word_in_sentence_has_index) + ")")

#line_index_with_span_for_lookup
word_index_for_lookup

[(0, 0, 'fire', 'NNP'),
 (1, 0, 'machinery', 'NN'),
 (2, 0, 'electric', 'JJ'),
 (3, 0, 'installation', 'NNS'),
 (4, 1, 'scope', 'NNP'),
 (5, 1, 'application', 'NN'),
 (6, 1, 'chapter', 'NN'),
 (7, 1, 'chapter', 'NN'),
 (8, 1, 'applies', 'NNS'),
 (9, 1, 'vessel', 'NNS'),
 (10, 1, 'construct', 'VBD'),
 (11, 2, 'vessel', 'NNS'),
 (12, 2, 'construct', 'VBN'),
 (13, 2, 'construct', 'VBN'),
 (14, 2, 'equipped', 'JJ'),
 (15, 2, 'accordance', 'NN'),
 (16, 2, 'rule', 'NNS'),
 (17, 2, 'applicable', 'JJ'),
 (18, 2, 'time', 'NN'),
 (19, 2, 'construction', 'NN'),
 (20, 2, 'vessel', 'NN'),
 (21, 2, 'otherwise', 'RB'),
 (22, 2, 'provide', 'VBN'),
 (23, 2, 'individual', 'JJ'),
 (24, 2, 'provision', 'NNS'),
 (25, 2, 'chapter', 'NN'),
 (26, 3, 'vessel', 'NN'),
 (27, 3, 'substantially', 'RB'),
 (28, 3, 'alter', 'VBD'),
 (29, 3, 'modified', 'JJ'),
 (30, 3, 'vessel', 'NN'),
 (31, 3, 'comply', 'VB'),
 (32, 3, 'requirement', 'NNS'),
 (33, 3, 'new', 'JJ'),
 (34, 3, 'vessel', 'NN'),
 (35, 4, 'amend', 'VBN'),
 

In [18]:
#
# TODO: Create bigrams
#

span_count = -1
line_index_with_span_for_lookup = []
word_index_for_lookup = []

# for each sentence in text
for line_idx,line in enumerate(lines_with_lemmatized_words_and_tag):
    
    number_of_words_in_line = len(line)
    first_word_in_sentence_has_index = span_count + 1
    last_word_in_sentence_has_index = first_word_in_sentence_has_index + number_of_words_in_line - 1
    
    # update span counter
    span_count = last_word_in_sentence_has_index
    
    line_index_with_span_for_lookup.append((line_idx,first_word_in_sentence_has_index,last_word_in_sentence_has_index))
    
    word_index_count = first_word_in_sentence_has_index
    
    sentence_as_list = []
    
    for (word_in_text,pos) in line:
        
        word_with_meta_info = (word_index_count,line_idx,word_in_text,pos)
        word_index_for_lookup.append(word_with_meta_info)
        
        sentence_as_list.append(word_in_text)
        
        # update counter
        word_index_count = word_index_count + 1
    
    sentence_as_text = " ".join(sentence_as_list)
    
    print(str(line_idx) + ": " + sentence_as_text)
    #print("Line index: " + str(line_idx) + " (" + str(first_word_in_sentence_has_index) + " - " + str(last_word_in_sentence_has_index) + ")")

#line_index_with_span_for_lookup
#word_index_for_lookup

0: fire machinery electric installation
1: scope application chapter chapter applies vessel construct
2: vessel construct construct equipped accordance rule applicable time construction vessel otherwise provide individual provision chapter
3: vessel substantially alter modified vessel comply requirement new vessel
4: amend regulation force
5: fire safety vessel construct structural fire protection accordance chapter nordic boat standard otherwise provide individual provision chapter
6: vessel construct structural fire protection accordance rule applicable time construction vessel
7: irrespective year construction vessel gasfired installation satisfy requirement regulation
8: safety precaution gasfired installation
9: operate propane liquefy hydrocarbon gas use board vessel
10: oil-fired stove stoves electric power supply oil supply shut automatically stove over-heated maximum °c smoke gas flame go failure occurs supply air power
11: stove electric power supply oil supply shut automatic

In [15]:
#
# Print out a selected sentence by the sentence index number
#

sentence = []

for (index_on_word,index_on_sentence,word,pos) in word_index_for_lookup:
    if index_on_sentence == 41:
        
        sentence.append(str(index_on_word) + " - " + word)
        
" ".join(sentence)
list(sentence)

['381 - smoke',
 '382 - uptake',
 '383 - carry',
 '384 - least',
 '385 - metre',
 '386 - deck']

In [16]:
#
# Create custom bigrams of sentences <-- TODO: berre sjå på kva du gjorde på TWO WORDS scriptet
#

#if index_of_line not in dictionary_with_known_data_properties_in_a_sentence:

from collections import defaultdict

# the dictionary index here is the sentence index number
sentences_for_finding_multiwords = defaultdict(int)

for (index_on_word,index_on_sentence,word,pos) in word_index_for_lookup:
    
    insert_word = (index_on_word,word,pos)
    
    if index_on_sentence not in sentences_for_finding_multiwords:
        sentences_for_finding_multiwords[index_on_sentence] = [insert_word]
    else:
        list_of_words = sentences_for_finding_multiwords[index_on_sentence]
        list_of_words.append(insert_word)
        sentences_for_finding_multiwords[index_on_sentence] = list_of_words
    
sentences_for_finding_multiwords

defaultdict(int,
            {0: [(0, 'fire', 'NNP'),
              (1, 'machinery', 'NN'),
              (2, 'electric', 'JJ'),
              (3, 'installation', 'NNS')],
             1: [(4, 'scope', 'NNP'),
              (5, 'application', 'NN'),
              (6, 'chapter', 'NN'),
              (7, 'chapter', 'NN'),
              (8, 'applies', 'NNS'),
              (9, 'vessel', 'NNS'),
              (10, 'construct', 'VBD')],
             2: [(11, 'vessel', 'NNS'),
              (12, 'construct', 'VBN'),
              (13, 'construct', 'VBN'),
              (14, 'equipped', 'JJ'),
              (15, 'accordance', 'NN'),
              (16, 'rule', 'NNS'),
              (17, 'applicable', 'JJ'),
              (18, 'time', 'NN'),
              (19, 'construction', 'NN'),
              (20, 'vessel', 'NN'),
              (21, 'otherwise', 'RB'),
              (22, 'provide', 'VBN'),
              (23, 'individual', 'JJ'),
              (24, 'provision', 'NNS'),
              (25, 'ch

In [ ]:
# create bigrams

for sentence_index_number in sentences_for_finding_multiwords:
    print(sentence_index_number)
    #print(sentences_for_finding_multiwords[sentence_index_number])
    
    #sentence_with_meta_in_a_list = sentences_for_finding_multiwords[sentence_index_number]
    #print(sentence_with_meta_in_a_list)
    
    #sentence_words_in_a_list = []
    
    #for (index_on_word,word,pos) in sentences_for_finding_multiwords[sentence_index_number]:
        #sentence_with_meta_in_a_list.append(what)
        #print(word)
        
    #print(sentence_words_in_a_list)
        
        
    #print(" ".join(sentence_words_in_a_list))




#sentence_words_in_a_list

In [21]:
#
# Hypotesis: We must know what to look for.
# Here is what we know.
#

# What are the things we know?
list_of_known_things = ['vessel', 'emergency light', 'pipe system', 'pipe', 'fire extinguisher']

# What are the data properties we know?
list_of_known_data_properties = ['length', 'overall length', 'closer', 'place closer', 'constructed after', 'constructed before', 'least']

# What are the data types we know?
list_of_known_data_types = ['cm', 'metre', 'voltage', 'mm']

# The big question: Do we find any of these three variables in the same sentence?
# Let's find out.

# Idea: Look for all the things we know.
# Next, in a selected line,
# if a data type is identified, 
# but at known data property is unknown, 
# then make a human identify the unknown data property.

In [22]:
#
# Find known data types in line
#

known_data_types_in_line = []

for word_we_know in list_of_known_data_types:
    
    # result
    # word_we_know -> data type example: cm
    # get sentence index
    index_of_lines_with_a_match = []
    
    for (word_index_number,line_index_number,word_in_text,pos) in word_index_for_lookup:

        if word_in_text == word_we_know:
            index_of_lines_with_a_match.append(line_index_number)
            #print(word_in_text + " " + str(line_index_number))
            
    if index_of_lines_with_a_match:
        known_data_types_in_line.append((word_we_know, index_of_lines_with_a_match))

known_data_types_in_line

[('cm', [38, 42, 44, 45, 48, 49, 49, 52, 89]),
 ('metre', [35, 41, 46, 64, 70, 76, 77, 105, 106, 107, 109, 112]),
 ('voltage', [105, 105, 106, 106, 107, 107, 109, 109, 113]),
 ('mm', [40])]

In [23]:
#
# Find known data properties in line
#

known_data_properties_in_line = []

for word_we_know in list_of_known_data_properties:
    
    index_of_lines_with_a_match = []
    
    for (word_index_number,line_index_number,word_in_text,pos) in word_index_for_lookup:

        if word_in_text == word_we_know:
            index_of_lines_with_a_match.append(line_index_number)
            #print(word_in_text + " " + str(line_index_number))
            
    if index_of_lines_with_a_match:
        known_data_properties_in_line.append((word_we_know, index_of_lines_with_a_match))

known_data_properties_in_line

[('length', [35, 64, 70, 76, 77, 105, 106, 107, 109, 112]),
 ('closer', [48, 52]),
 ('least',
  [38,
   40,
   41,
   42,
   44,
   45,
   46,
   49,
   77,
   77,
   77,
   78,
   78,
   79,
   79,
   83,
   88,
   89,
   89,
   112])]

In [24]:
# List the findings

findings = []

for (word_we_know, index_of_lines_with_a_match) in known_data_types_in_line:
    for index_of_line in index_of_lines_with_a_match:
        findings.append((index_of_line,word_we_know))

sortedList = sorted(findings, key = lambda y: (y[0]))
sortedList

[(35, 'metre'),
 (38, 'cm'),
 (40, 'mm'),
 (41, 'metre'),
 (42, 'cm'),
 (44, 'cm'),
 (45, 'cm'),
 (46, 'metre'),
 (48, 'cm'),
 (49, 'cm'),
 (49, 'cm'),
 (52, 'cm'),
 (64, 'metre'),
 (70, 'metre'),
 (76, 'metre'),
 (77, 'metre'),
 (89, 'cm'),
 (105, 'metre'),
 (105, 'voltage'),
 (105, 'voltage'),
 (106, 'metre'),
 (106, 'voltage'),
 (106, 'voltage'),
 (107, 'metre'),
 (107, 'voltage'),
 (107, 'voltage'),
 (109, 'metre'),
 (109, 'voltage'),
 (109, 'voltage'),
 (112, 'metre'),
 (113, 'voltage')]

In [25]:
# Create dictionary with findings of known_data_types_in_line
# The dictionary key is the index number of the line

from collections import defaultdict

dictionary_with_known_data_types_in_a_sentence = defaultdict(int)

for (word_we_know, index_of_lines_with_a_match) in known_data_types_in_line:
    
    for index_of_line in index_of_lines_with_a_match:
        
        if index_of_line not in dictionary_with_known_data_types_in_a_sentence:
            
            print('First detection in sentence ' + str(index_of_line) + " - " + word_we_know)
            
            # Add to dictionary
            dictionary_with_known_data_types_in_a_sentence[index_of_line] = [word_we_know]
            
        else: # a detection in sentence is already added
            
            print('New detection in sentence ' + str(index_of_line) + " - " + word_we_know)
            
            # Add if it is a new word
            if word_we_know not in dictionary_with_known_data_types_in_a_sentence[index_of_line]:
                dictionary_with_known_data_types_in_a_sentence[index_of_line].append(word_we_know)
                
dictionary_with_known_data_types_in_a_sentence

First detection in sentence 38 - cm
First detection in sentence 42 - cm
First detection in sentence 44 - cm
First detection in sentence 45 - cm
First detection in sentence 48 - cm
First detection in sentence 49 - cm
New detection in sentence 49 - cm
First detection in sentence 52 - cm
First detection in sentence 89 - cm
First detection in sentence 35 - metre
First detection in sentence 41 - metre
First detection in sentence 46 - metre
First detection in sentence 64 - metre
First detection in sentence 70 - metre
First detection in sentence 76 - metre
First detection in sentence 77 - metre
First detection in sentence 105 - metre
First detection in sentence 106 - metre
First detection in sentence 107 - metre
First detection in sentence 109 - metre
First detection in sentence 112 - metre
New detection in sentence 105 - voltage
New detection in sentence 105 - voltage
New detection in sentence 106 - voltage
New detection in sentence 106 - voltage
New detection in sentence 107 - voltage
New d

defaultdict(int,
            {38: ['cm'],
             42: ['cm'],
             44: ['cm'],
             45: ['cm'],
             48: ['cm'],
             49: ['cm'],
             52: ['cm'],
             89: ['cm'],
             35: ['metre'],
             41: ['metre'],
             46: ['metre'],
             64: ['metre'],
             70: ['metre'],
             76: ['metre'],
             77: ['metre'],
             105: ['metre', 'voltage'],
             106: ['metre', 'voltage'],
             107: ['metre', 'voltage'],
             109: ['metre', 'voltage'],
             112: ['metre'],
             113: ['voltage'],
             40: ['mm']})

In [26]:
# Create dictionary with findings of known_data_properties_in_line
# The dictionary key is the index number of the line

from collections import defaultdict

dictionary_with_known_data_properties_in_a_sentence = defaultdict(int)

for (word_we_know, index_of_lines_with_a_match) in known_data_properties_in_line:
    
    for index_of_line in index_of_lines_with_a_match:
        
        if index_of_line not in dictionary_with_known_data_properties_in_a_sentence:
            
            print('First detection in sentence ' + str(index_of_line) + " - " + word_we_know)
            
            # Add to dictionary
            dictionary_with_known_data_properties_in_a_sentence[index_of_line] = [word_we_know]
            
        else: # a detection in sentence is already added
            
            print('New detection in sentence ' + str(index_of_line) + " - " + word_we_know)
            
            # Add if it is a new word
            if word_we_know not in dictionary_with_known_data_properties_in_a_sentence[index_of_line]:
                dictionary_with_known_data_properties_in_a_sentence[index_of_line].append(word_we_know)
                
dictionary_with_known_data_properties_in_a_sentence

First detection in sentence 35 - length
First detection in sentence 64 - length
First detection in sentence 70 - length
First detection in sentence 76 - length
First detection in sentence 77 - length
First detection in sentence 105 - length
First detection in sentence 106 - length
First detection in sentence 107 - length
First detection in sentence 109 - length
First detection in sentence 112 - length
First detection in sentence 48 - closer
First detection in sentence 52 - closer
First detection in sentence 38 - least
First detection in sentence 40 - least
First detection in sentence 41 - least
First detection in sentence 42 - least
First detection in sentence 44 - least
First detection in sentence 45 - least
First detection in sentence 46 - least
First detection in sentence 49 - least
New detection in sentence 77 - least
New detection in sentence 77 - least
New detection in sentence 77 - least
First detection in sentence 78 - least
New detection in sentence 78 - least
First detection 

defaultdict(int,
            {35: ['length'],
             64: ['length'],
             70: ['length'],
             76: ['length'],
             77: ['length', 'least'],
             105: ['length'],
             106: ['length'],
             107: ['length'],
             109: ['length'],
             112: ['length', 'least'],
             48: ['closer'],
             52: ['closer'],
             38: ['least'],
             40: ['least'],
             41: ['least'],
             42: ['least'],
             44: ['least'],
             45: ['least'],
             46: ['least'],
             49: ['least'],
             78: ['least'],
             79: ['least'],
             83: ['least'],
             88: ['least'],
             89: ['least']})

In [27]:
# Merge dictionary_with_known_data_types_in_a_sentence and dictionary_with_known_data_properties_in_a_sentence

dictionary_with_known_data_types_and_properties_in_a_sentence = defaultdict(int)

# Load data type line numbers
for line_number in dictionary_with_known_data_types_in_a_sentence.keys():
    dictionary_with_known_data_types_and_properties_in_a_sentence[line_number] = ([],[])

# Load data property line numbers
for line_number in dictionary_with_known_data_properties_in_a_sentence.keys():
    dictionary_with_known_data_types_and_properties_in_a_sentence[line_number] = ([],[])

# Load data
for line_number in dictionary_with_known_data_types_and_properties_in_a_sentence.keys():
    data_types = dictionary_with_known_data_types_in_a_sentence[line_number]
    data_properties = dictionary_with_known_data_properties_in_a_sentence[line_number]    
    dictionary_with_known_data_types_and_properties_in_a_sentence[line_number] = (data_types,data_properties)

# Result comment: easy to see what we have.
# A human must look at the sentences with 0 values
    
dictionary_with_known_data_types_and_properties_in_a_sentence

defaultdict(int,
            {38: (['cm'], ['least']),
             42: (['cm'], ['least']),
             44: (['cm'], ['least']),
             45: (['cm'], ['least']),
             48: (['cm'], ['closer']),
             49: (['cm'], ['least']),
             52: (['cm'], ['closer']),
             89: (['cm'], ['least']),
             35: (['metre'], ['length']),
             41: (['metre'], ['least']),
             46: (['metre'], ['least']),
             64: (['metre'], ['length']),
             70: (['metre'], ['length']),
             76: (['metre'], ['length']),
             77: (['metre'], ['length', 'least']),
             105: (['metre', 'voltage'], ['length']),
             106: (['metre', 'voltage'], ['length']),
             107: (['metre', 'voltage'], ['length']),
             109: (['metre', 'voltage'], ['length']),
             112: (['metre'], ['length', 'least']),
             113: (['voltage'], 0),
             40: (['mm'], ['least']),
             78: (0, ['least']),
 

In [28]:
# Focus on sentences with findings on both type AND property

dictionary_with_match_on_type_and_property = defaultdict(int)
dictionary_with_match_on_type_or_property = defaultdict(int)

for line_number in dictionary_with_known_data_types_and_properties_in_a_sentence.keys():
    data_types = dictionary_with_known_data_types_and_properties_in_a_sentence[line_number][0]
    data_properties = dictionary_with_known_data_types_and_properties_in_a_sentence[line_number][1]
    
    if data_types == 0 or data_properties == 0:
        dictionary_with_match_on_type_or_property[line_number] = (data_types,data_properties)
    else:
        dictionary_with_match_on_type_and_property[line_number] = (data_types,data_properties)

dictionary_with_match_on_type_and_property
#dictionary_with_match_on_type_or_property

defaultdict(int,
            {38: (['cm'], ['least']),
             42: (['cm'], ['least']),
             44: (['cm'], ['least']),
             45: (['cm'], ['least']),
             48: (['cm'], ['closer']),
             49: (['cm'], ['least']),
             52: (['cm'], ['closer']),
             89: (['cm'], ['least']),
             35: (['metre'], ['length']),
             41: (['metre'], ['least']),
             46: (['metre'], ['least']),
             64: (['metre'], ['length']),
             70: (['metre'], ['length']),
             76: (['metre'], ['length']),
             77: (['metre'], ['length', 'least']),
             105: (['metre', 'voltage'], ['length']),
             106: (['metre', 'voltage'], ['length']),
             107: (['metre', 'voltage'], ['length']),
             109: (['metre', 'voltage'], ['length']),
             112: (['metre'], ['length', 'least']),
             40: (['mm'], ['least'])})

In [29]:
#
# Here we have found sentences with something, but type OR property is missing.
# This is a good place for a human to look for sentences with unknown words.
#

dictionary_with_match_on_type_and_property = defaultdict(int)
dictionary_with_match_on_type_or_property = defaultdict(int)

for line_number in dictionary_with_known_data_types_and_properties_in_a_sentence.keys():
    data_types = dictionary_with_known_data_types_and_properties_in_a_sentence[line_number][0]
    data_properties = dictionary_with_known_data_types_and_properties_in_a_sentence[line_number][1]
    
    if data_types == 0 or data_properties == 0:
        dictionary_with_match_on_type_or_property[line_number] = (data_types,data_properties)
    else:
        dictionary_with_match_on_type_and_property[line_number] = (data_types,data_properties)

#dictionary_with_match_on_type_and_property
dictionary_with_match_on_type_or_property

defaultdict(int,
            {113: (['voltage'], 0),
             78: (0, ['least']),
             79: (0, ['least']),
             83: (0, ['least']),
             88: (0, ['least'])})

# Identify known and unknown things in sentences

In [30]:
#
# Find known words in line
#
# Now we have identified lines with known types and properties.
# Next up, can we identify the thing we are talking about?
#
# Idea: Remove the identified words from the sentences before continuing to searching in them

selected_line_in_a_list_of_words = []

for (word_index_number,line_index_number,word_in_text,pos) in word_index_for_lookup:
    if line_index_number == 105:
        selected_line_in_a_list_of_words.append((word_index_number,line_index_number,word_in_text,pos))

line_with_findings_removed = [(word_index_number,line_index_number,word_in_text,pos) for (word_index_number,line_index_number,word_in_text,pos) in selected_line_in_a_list_of_words if word_in_text not in list_of_known_data_types]
line_with_findings_removed

[(1047, 105, 'electrical', 'JJ'),
 (1048, 105, 'installation', 'NNS'),
 (1049, 105, 'vessel', 'NNP'),
 (1051, 105, 'overall', 'JJ'),
 (1052, 105, 'length', 'NN'),
 (1053, 105, 'construct', 'VBD'),
 (1054, 105, 'electrical', 'JJ'),
 (1055, 105, 'installation', 'NNS'),
 (1058, 105, 'satisfy', 'VB'),
 (1059, 105, 'requirement', 'NNS'),
 (1060, 105, 'nordic', 'NNP'),
 (1061, 105, 'boat', 'NNP'),
 (1062, 105, 'standard', 'NNP'),
 (1063, 105, 'electrical', 'JJ'),
 (1064, 105, 'installation', 'NNS')]

In [31]:
#
# Find sentences where we are going to look for something (the subject in the RDf triple)
#
# Now we have identified lines with known types and properties.
# Next up, can we identify the thing we are talking about?
#
# Idea: Remove the identified words from the sentences before continuing to searching in them

sentences_without_data_type_and_data_property = []

for line_number in dictionary_with_match_on_type_and_property.keys():
    for (word_index_number,line_index_number,word_in_text,pos) in word_index_for_lookup:
        
        # only add words from sentence that is not data_type or data_property
        data_types = dictionary_with_match_on_type_and_property[line_number][0]
        data_properties = dictionary_with_match_on_type_and_property[line_number][1]

        if line_index_number == line_number and word_in_text not in data_types and word_in_text not in data_properties:
            sentences_without_data_type_and_data_property.append((word_index_number,line_index_number,word_in_text,pos))

sentences_without_data_type_and_data_property

[(336, 38, 'stove', 'NNS'),
 (337, 38, 'wooden', 'JJ'),
 (338, 38, 'floor', 'NN'),
 (339, 38, 'combustible', 'JJ'),
 (340, 38, 'floor', 'NN'),
 (341, 38, 'cover', 'VBG'),
 (342, 38, 'protect', 'VBN'),
 (343, 38, 'steel', 'NN'),
 (344, 38, 'aluminium', 'NN'),
 (345, 38, 'sheet', 'NNS'),
 (346, 38, 'cover', 'VBG'),
 (347, 38, 'whole', 'JJ'),
 (348, 38, 'floor', 'NN'),
 (349, 38, 'area', 'NN'),
 (350, 38, 'fireplace', 'NN'),
 (351, 38, 'front', 'JJ'),
 (352, 38, 'stove', 'NN'),
 (355, 38, 'fire', 'NN'),
 (356, 38, 'door', 'NN'),
 (387, 42, 'saucer', 'NNP'),
 (388, 42, 'form', 'VBD'),
 (389, 42, 'rockwool', 'JJ'),
 (390, 42, 'sheet', 'NNS'),
 (393, 42, 'thickness', 'NN'),
 (394, 42, 'provide', 'VB'),
 (395, 42, 'smoke', 'NN'),
 (396, 42, 'uptake', 'JJ'),
 (397, 42, 'combustible', 'JJ'),
 (398, 42, 'material', 'NN'),
 (409, 44, 'alternatively', 'RB'),
 (410, 44, 'pipe', 'JJ'),
 (411, 44, 'diameter', 'NN'),
 (414, 44, 'large', 'JJR'),
 (415, 44, 'smoke', 'NN'),
 (416, 44, 'uptake', 'VBP'),
 

## Find single word suggestions

In [32]:
test = [(1057, 105, 'electrical', 'JJ'),
 (1058, 105, 'installation', 'NNS'),
 (1059, 105, 'vessel', 'NNP'),
 (1061, 105, 'overall', 'JJ'),
 (1063, 105, 'construct', 'VBD'),
 (1065, 105, 'electrical', 'JJ'),
 (1066, 105, 'installation', 'NNS'),
 (1069, 105, 'satisfy', 'VB'),
 (1070, 105, 'requirement', 'NNS'),
 (1071, 105, 'nordic', 'NNP'),
 (1072, 105, 'boat', 'NNP'),
 (1073, 105, 'standard', 'NNP'),
 (1074, 105, 'electrical', 'JJ'),
 (1075, 105, 'installation', 'NNS')]

# her ser ein eit mønster 1071 + 1072 + 1073 == nordic boat standard

# TODO: create CSV with suggestions of things of new things

# TODO: suggest things we know, that we find, list_of_known_things = ['vessel', 'emergency light', 'pipe system', 'pipe', 'fire extinguisher']

for (word_index_number,line_index_number,word_in_text,pos) in test:
    if pos.startswith('NNP'):
        print((word_index_number,line_index_number,word_in_text,pos))

(1059, 105, 'vessel', 'NNP')
(1071, 105, 'nordic', 'NNP')
(1072, 105, 'boat', 'NNP')
(1073, 105, 'standard', 'NNP')


In [33]:
# print the nouns
for line_number in dictionary_with_match_on_type_and_property.keys():
    for (word_index_number,line_index_number,word_in_text,pos) in sentences_without_data_type_and_data_property:
        if line_number == line_index_number and pos.startswith('N'):
            print((word_index_number,line_index_number,word_in_text,pos))

(336, 38, 'stove', 'NNS')
(338, 38, 'floor', 'NN')
(340, 38, 'floor', 'NN')
(343, 38, 'steel', 'NN')
(344, 38, 'aluminium', 'NN')
(345, 38, 'sheet', 'NNS')
(348, 38, 'floor', 'NN')
(349, 38, 'area', 'NN')
(350, 38, 'fireplace', 'NN')
(352, 38, 'stove', 'NN')
(355, 38, 'fire', 'NN')
(356, 38, 'door', 'NN')
(387, 42, 'saucer', 'NNP')
(390, 42, 'sheet', 'NNS')
(393, 42, 'thickness', 'NN')
(395, 42, 'smoke', 'NN')
(398, 42, 'material', 'NN')
(411, 44, 'diameter', 'NN')
(415, 44, 'smoke', 'NN')
(417, 44, 'pipe', 'NN')
(420, 44, 'order', 'NN')
(423, 44, 'air', 'NN')
(424, 44, 'circulation', 'NN')
(425, 44, 'pipe', 'NNS')
(426, 45, 'insulation', 'NNP')
(427, 45, 'saucer', 'NN')
(431, 45, 'wool', 'NN')
(432, 45, 'twine', 'NN')
(435, 45, 'thickness', 'NN')
(437, 45, 'outer', 'NN')
(438, 45, 'pipe', 'NN')
(440, 45, 'material', 'NN')
(457, 48, 'smoke', 'NNP')
(458, 48, 'uptake', 'NNS')
(463, 48, 'bulkhead', 'NNS')
(464, 49, 'smoke', 'NN')
(466, 49, 'pipe', 'NN')
(468, 49, 'saucer', 'NN')
(472, 49

In [34]:
# TODO: compare to list of known things and print out to csv
# TODO: do this for 1, 2, 3 and 4 word combinations
# TODO: check words for repeating words and suggest thouse
# TODO: create fragments of a sentence, and create bi and thrigrams of thouse fragments

In [35]:
# Look for words that we know

for line_number in dictionary_with_match_on_type_and_property.keys():
    for (word_index_number,line_index_number,word_in_text,pos) in sentences_without_data_type_and_data_property:
        if line_number == line_index_number and pos.startswith('N'):
            if word_in_text in list_of_known_things:
                print((word_index_number,line_index_number,word_in_text,pos))

(417, 44, 'pipe', 'NN')
(425, 44, 'pipe', 'NNS')
(438, 45, 'pipe', 'NN')
(466, 49, 'pipe', 'NN')
(498, 52, 'pipe', 'NNS')
(318, 35, 'vessel', 'NNS')
(441, 46, 'pipe', 'NN')
(630, 64, 'vessel', 'NNP')
(674, 70, 'vessel', 'NNP')
(742, 76, 'vessel', 'NNP')
(763, 77, 'vessel', 'NNP')
(1049, 105, 'vessel', 'NNP')
(1065, 106, 'vessel', 'NNS')
(1079, 107, 'vessel', 'NNS')
(1097, 109, 'vessel', 'NNS')
(1124, 112, 'vessel', 'NNS')
(368, 40, 'pipe', 'NNS')
(372, 40, 'pipe', 'NNS')


Comment on the result:
Are we only alking about "vessel" here?
It is the only recognised word in our test vocabulary / ontology

(740, 76, 'vessel', 'NNP')
(1047, 105, 'vessel', 'NNP')
(1063, 106, 'vessel', 'NNS')
(1077, 107, 'vessel', 'NNS')
(1095, 109, 'vessel', 'NNS')

-- Update: Added 'least' to list_of_known_data_properties. New result:

(416, 44, 'pipe', 'NN')
(424, 44, 'pipe', 'NNS')
(437, 45, 'pipe', 'NN')
(464, 49, 'pipe', 'NN')
(740, 76, 'vessel', 'NNP')
(1047, 105, 'vessel', 'NNP')
(1063, 106, 'vessel', 'NNS')
(1077, 107, 'vessel', 'NNS')
(1095, 109, 'vessel', 'NNS')
(368, 40, 'pipe', 'NNS')
(372, 40, 'pipe', 'NNS')

Here we can see that 'pipe' is identified as something we know.


 -- Update: Added custom lemmatizer on 'metres' to 'metre'
 
 (416, 44, 'pipe', 'NN')
(424, 44, 'pipe', 'NNS')
(437, 45, 'pipe', 'NN')
(464, 49, 'pipe', 'NN')
(318, 35, 'vessel', 'NNS')
(628, 64, 'vessel', 'NNP')
(672, 70, 'vessel', 'NNP')
(1063, 106, 'vessel', 'NNS')
(1077, 107, 'vessel', 'NNS')
(1095, 109, 'vessel', 'NNS')
(1122, 112, 'vessel', 'NNS')
(1047, 105, 'vessel', 'NNP')
(368, 40, 'pipe', 'NNS')
(372, 40, 'pipe', 'NNS')

In [36]:
# Look for words that we know

test = []

for line_number in dictionary_with_match_on_type_and_property.keys():
    for (word_index_number,line_index_number,word_in_text,pos) in sentences_without_data_type_and_data_property:
        if line_number == line_index_number and pos.startswith('N'):
            if word_in_text in list_of_known_things:
                data_types = dictionary_with_match_on_type_and_property[line_number][0]
                data_properties = dictionary_with_match_on_type_and_property[line_number][1]
                
                for data_type in data_types:
                    
                    for data_property in data_properties:
                        
                        test.append(word_in_text + "," + data_property + "," + data_type)

test

['pipe,least,cm',
 'pipe,least,cm',
 'pipe,least,cm',
 'pipe,least,cm',
 'pipe,closer,cm',
 'vessel,length,metre',
 'pipe,least,metre',
 'vessel,length,metre',
 'vessel,length,metre',
 'vessel,length,metre',
 'vessel,length,metre',
 'vessel,least,metre',
 'vessel,length,metre',
 'vessel,length,voltage',
 'vessel,length,metre',
 'vessel,length,voltage',
 'vessel,length,metre',
 'vessel,length,voltage',
 'vessel,length,metre',
 'vessel,length,voltage',
 'vessel,length,metre',
 'vessel,least,metre',
 'pipe,least,mm',
 'pipe,least,mm']

Comment on result

['vessel,length,metres',
 'vessel,length,metres',
 'vessel,length,v',
 'vessel,length,v',
 'vessel,length,v',
 'vessel,length,v']
 
 We are only talking about one thing here, and we have a lot of duplicates form the different sentences.
 
-- Update: Added 'least' to list_of_known_data_properties. New result:
 
 ['pipe,least,cm',
 'pipe,least,cm',
 'pipe,least,cm',
 'pipe,least,cm',
 'vessel,length,metres',
 'vessel,length,metres',
 'vessel,length,v',
 'vessel,length,v',
 'vessel,length,v',
 'vessel,length,v',
 'pipe,least,mm',
 'pipe,least,mm']
 
 -- Update: Added custom lemmatizer on 'metres' to 'metre'
 
 ['pipe,least,cm',
 'pipe,least,cm',
 'pipe,least,cm',
 'pipe,least,cm',
 'vessel,length,metre',
 'vessel,length,metre',
 'vessel,length,metre',
 'vessel,length,metre',
 'vessel,length,v',
 'vessel,length,metre',
 'vessel,length,v',
 'vessel,length,metre',
 'vessel,length,v',
 'vessel,length,metre',
 'vessel,least,metre',
 'vessel,length,v',
 'pipe,least,mm',
 'pipe,least,mm']
 
 TODO: Length and voltage do not belong together

In [37]:
# Removing duplicates from result and make CSV output ready

output_without_duplicates = list(dict.fromkeys(test))
lines_for_csv_file = []

for line in output_without_duplicates:
    line_split = line.split(",")
    lines_for_csv_file.append((line_split[0],line_split[1],line_split[2]))

lines_for_csv_file

[('pipe', 'least', 'cm'),
 ('pipe', 'closer', 'cm'),
 ('vessel', 'length', 'metre'),
 ('pipe', 'least', 'metre'),
 ('vessel', 'least', 'metre'),
 ('vessel', 'length', 'voltage'),
 ('pipe', 'least', 'mm')]

In [38]:
import csv

with open("output - Identify data type properties - Multi words.csv", 'w') as f:
    write = csv.writer(f, delimiter=',', lineterminator='\n')
    write.writerows(lines_for_csv_file)